solidity/contracts/bridge/Wallets.sol

```diff
@@ -456,6 +464,115 @@ library Wallets {
         self.ecdsaWalletRegistry.closeWallet(wallet.ecdsaWalletID);
     }
 
+    /// @notice Notifies about the submission of the moving funds target
+    ///         wallets commitment. Once all requirements are met, that
+    ///         function registers the target wallets commitment and opens
+    ///         the way for moving funds proof submission.
+    /// @param walletPubKeyHash 20-byte public key hash of the source wallet
+    /// @param walletMainUtxo Data of the source wallet's main UTXO, as
+    ///        currently known on the Ethereum chain.
+    /// @param walletMembersIDs Identifiers of the source wallet signing group
+    ///        members
+    /// @param walletMemberIndex Position of the caller in the source wallet
+    ///        signing group members list
+    /// @param targetWallets List of 20-byte public key hashes of the target
+    ///        wallets that the source wallet commits to
+    /// @dev Requirements:
+    ///      - The source wallet must be in the MovingFunds state
+    ///      - The source wallet must not have pending redemption requests
+    ///      - The source wallet must not have prior commitments submitted earlier
+    ///      - The expression `keccak256(abi.encode(walletMembersIDs))` must
+    ///        be exactly the same as the hash stored under `membersIdsHash`
+    ///        for the given source wallet in the ECDSA registry. Those IDs are
+    ///        not directly stored in the contract for gas efficiency purposes
+    ///        but they can be read from appropriate `DkgResultSubmitted`
+    ///        and `DkgResultApproved` events.
+    ///      - The `walletMemberIndex` must be in range [1, walletMembersIDs.length]
+    ///      - The caller must be the member of the source wallet signing group
+    ///        at the position indicated by `walletMemberIndex` parameter
+    ///      - The `walletMainUtxo` components must point to the recent main
+    ///        UTXO of the source wallet, as currently known on the Ethereum
+    ///        chain. If the source wallet has no main UTXO, this parameter
+    ///        can be empty as it is ignored since the wallet balance is
+    ///        assumed to be zero.
+    ///      - At least one Live wallet must exist in the system
+    ///      - Submitted target wallets count must match the expected count
+    ///        `N = min(liveWalletsCount, ceil(walletBtcBalance / walletMaxBtcTransfer))`
+    ///        where `N > 0`
+    ///      - Each target wallet must be not equal to the source wallet
+    ///      - Each target wallet must be in Live state
+    function notifyWalletMovingFundsCommitmentSubmitted(
```

Comment:
Why in `Wallets.sol` and not in `MovingFunds.sol`? 🤔

Reply:
The responsibility boundaries between those two are very fluid 😀 But, this is how I see it:
- `Wallets` should be responsible for wallet state updates (ideally)
- `MovingFunds` library is here because we needed to reduce the size of the `Bridge`. It groups everything related to moving funds that would normally live in the `Bridge`. Giving the above point about `Wallets`, the `MovingFunds` library should delegate wallet management to `Wallets`. A good example is moving funds proof where BTC tx processing is done in `MovingFunds` and wallet state check and update is made in `Wallets`.

Moving funds commitment is strictly related to wallet state updates. This is why `Wallets` library does the heavy lifting and `MovingFunds` just delegates the call. We can do it directly from `Bridge` but if we want to keep things consistent, that should pass through `MovingFunds` imho.

Reply:
> * `Wallets` should be responsible for wallet state updates (ideally)

Exactly! 💯 

But see that `notifyWalletMovingFundsCommitmentSubmitted` is not changing the state of the wallet. The wallet was in MovingFunds state and the wallet is still in MovingFunds state. It's only accepting information that will be later used by `MovingFunds.submitMovingFundsProof`. I think we should have this code in `MovingFunds` library. `MovingFunds.submitMovingFundsProof` is delegating to `Wallets.notifyWalletFundsMoved` because it's changing the state of the wallet and it makes perfect sense. `submitMovingFundsCommitment` is not changing it.

Reply:
Also, looking at it again, I think this part:
https://github.com/keep-network/tbtc-v2/blob/5c95ed5c36e440c9aca25ec4f2694f9d8966f84f/solidity/contracts/bridge/Wallets.sol#L490-L503

should be in `MovingFunds.submitMovingFundsProof`. All before-state validations, such as `wallet.state == WalletState.MovingFunds`, should stay in `Wallets.sol` IMO in `notify*` functions.

Reply:
> But see that notifyWalletMovingFundsCommitmentSubmitted is not changing the state of the wallet.

I think we have ambiguity when talking about "state". When I said:

> Wallets should be responsible for wallet state updates (ideally)

I meant that `Wallets` should take care of updates touching all fields in the `Wallet` struct, not just the `state` field. This is why I think it should NOT be done in `MovingFunds` because there is a more specific `Wallets` library that can manage everything about wallets. If we move those parts to `MovingFunds` we will scatter the wallet management logic and we will not have everything about wallets in one place as we aimed for.

Reply:
We are already having cases where fields from `Wallets` struct are updated outside of `Wallets` library, one example is `Redemption` and `Sweep` changing `mainUtxoHash` and I think this is fine. 

Updating `mainUtxoHash` from `Redemption` and `Sweep` library is natural. In the same way as updating `movingFundsTargetWalletsCommitmentHash` from `MovingFunds` library is natural for me.

Also, see that the fact we route this call through `MovingFunds` is a sign that this is `MovingFunds`-related activity. 

I think `MovingFunds` should be only for state transitions and wallet lifecycle between transitions. It needs to validate before & after state when doing the transition and `Redemption`, `Sweep`, `MovingFunds` should be for particular activities validating the state required to perform such an activity.

Reply:
> We are already having cases where fields from Wallets struct are updated outside of Wallets library, one example is Redemption and Sweep changing mainUtxoHash and I think this is fine.

Yeah, but we talked about refactoring it by exposing appropriate functions from `Wallets` library and avoiding direct manipulation of the wallet state. 

> Also, see that the fact we route this call through MovingFunds is a sign that this is MovingFunds-related activity.

Indeed it is. However, the commitment is attached strictly to the wallet state. This is something other than, for example, a deposit which is a separate being and lies in a part of the contract state managed strictly by the `Deposit` library. Also, `MovingFunds` is here just because moving funds consist of some code that process BTC transaction. Otherwise, all this stuff could live directly in `Wallets` library.

> I think MovingFunds should be only for state transitions and wallet lifecycle between transitions. It needs to validate before & after state when doing the transition and Redemption, Sweep, MovingFunds should be for particular activities validating the state required to perform such an activity.

Agreed but see, the commitment is just a substep of the transition from `MovingFunds` to `Closing/Closed`. The wallet cannot change the state if the commitment is wrong or not submitted at all.

Reply:
> Yeah, but we talked about refactoring it by exposing appropriate functions from Wallets library and avoiding direct manipulation of the wallet state.

How would such a function look like for Redemption/Sweeping though? Is it just about having an internal setter function in `Wallets.sol` for updating mainUTXO? If so, we can do the same for setting target wallets commitment 🤷🏻

Reply:
Not quite, that can be more complex than just simple getters/setters. Let's look at requesting redemption. We can expose `canAcceptRedemptionRequest` where we will check the wallet state is `Live` and the wallet has enough of free balance to handle that redemption. Regarding setting the main UTXO for sweeping/redemption, for the sweep, that's indeed a simple setter but for redemption, there is an additional logic that checks whether we should set or unset the main UTXO altogether.

Reply:
What if we decide about it at the very end, in a single PR, once we have all elements in place and for now stick to the current convention?

I am not saying I do not like this idea but to see if it is actually good or not, I would need to see the code. It may turn out that the code is easier to read but it may also turn out that there are too many calls to follow.

So far, the convention was that `Wallets.sol` is about requesting/registering new wallet and state transitions - checking the in and out state and performing the transition.

All validations specific to the given action were done in respective libraries. For example, the mentioned `Redemption.requestRedemption` checks the state of the wallet and its balance. In the end, it is modifying the wallet's main UTXO as something that is redemption/sweep-specific.

"Move funds" is a special case of redemption. Using this analogy, commitment for target wallets is like requesting redemption, and proving the movement of funds is like redemption proof. Using this analogy, I think the entire code of `notifyWalletMovingFundsCommitmentSubmitted` should be in `MovingFunds` and this part should be in `MovingFunds.submitMovingFundsProof`:
https://github.com/keep-network/tbtc-v2/blob/d8aabc13bf4bcc701ece6ff36e5c5eb7f1c441ef/solidity/contracts/bridge/Wallets.sol#L610-L623

Again - not saying I do not like it - but I am just worried that the final result may not be as clean as we think it's going to be and I would rather do it at the end once we have all the building blocks instead of having two competing conventions in the code now. 

I do not think the current convention is that bad, to be honest. When I am looking for a logic specific to the given action, I can see all the validations (requirements) in one place. And when I want to see if the wallet can transition from state x to state y, I can also see it all in one place with no cluttering code in between.

Reply:
Ok, there is no point in blocking too long here so let's move the commitment to `MovingFunds`. Just for the sake of potential future refactoring, I'll dump some concerns I have in my brain to have them captured somewhere:

The problem I see with the current convention is that it is not precise enough to be called a convention. This is the reason why we have this discussion at all.  The responsibility of `Wallets` library after the Bridge size-oriented refactoring is not clear and makes the navigation non-intuitive for someone not familiar with the code. We both have different points of view regarding that library and I bet another person will have a yet different opinion 😄 

In my opinion, the `Wallets` library must have one of the two possible shapes:
- We follow the rules about validating and updating the wallet's high-level state (everything living in `Wallet` struct) outside and the `Wallets` library only registers the new wallets and exposes methods to make state transitions: `moveFunds`, `terminateWallet`, `closeWallet`. The Bridge is responsible to decide how and when the state transitions should be made. This option is basically the extension of what we have now. That also means `Wallets` should not have any `notify*` functions exposed. 
- We put everything wallet-related into `Wallets` library, just as I proposed earlier. The Bridge just calls appropriate methods from `Wallets` library to validate and update wallet's high-level state.

What we have now is something between those two shapes and this is the main problem in my opinion.

Reply:
Makes sense!

Reply:
See https://github.com/keep-network/tbtc-v2/pull/208/commits/a740fb56349974c13cb3931510d5e238bd69e570